zebrad/src/components/mempool.rs

```diff
@@ -100,6 +146,38 @@ impl Service<Request> for Mempool {
                     .map(Response::RejectedTransactionIds);
                 async move { rsp }.boxed()
             }
+            Request::DownloadAndVerify(gossiped_txs) => {
+                let rsp = gossiped_txs
+                    .into_iter()
+                    .map(|gossiped_tx| {
+                        let r = self.should_download_or_verify(gossiped_tx.id());
+                        if let Err(action) = r {
+                            return action;
+                        }
+                        DownloadAction::DownloadAction(
+                            self.tx_downloads.download_if_needed_and_verify(gossiped_tx),
+                        )
+                    })
+                    .collect();
+                async move { Ok(Response::DownloadActions(rsp)) }.boxed()
+            }
+        }
+    }
+}
+
+impl Mempool {
+    /// Check if transaction should be downloaded and/or verified.
+    ///
+    /// If it is already in the mempool (or in its rejected list)
+    /// then it shouldn't be downloaded/verified.
+    fn should_download_or_verify(&mut self, txid: UnminedTxId) -> Result<(), DownloadAction> {
```

Comment:
I'm not sure if this should return an `Err(DownloadAction)`, or if it should just return the `DownloadAction` enum variant?

Reply:
Revamped in 8c29d130, I'm just using `MempoolError`, we can specialize if needed in #2668